### PR TITLE
feat(api): add host option in launchServer options

### LINKED ISSUE #30998: [Feature]: launchServer bind hostname or address
### 🚀 Feature Request

I really like the server method that playwright provides so that I can remotely control the test nodes. Currently, I can use docker to start a playwright server and expose ports to a control node playwright client for testing. In contrast, when I run playwright server directly without docker, I am only able to get the ws url of `localhost`, which prevents me from connecting to it directly through other control nodes within a connectable network.

### Example

```
const { chromium, webkit, firefox } = require('playwright');  // Or 'webkit' or 'firefox'.

(async () => {
  [chromium, webkit, firefox].forEach(async browserType => {
    const browserServer = await browserType.launchServer({
      // host: "0.0.0.0", // unsupported yet
      // port: 55432,
      // wsPath: "",
    });
    const wsEndpoint = browserServer.wsEndpoint();
    console.log(wsEndpoint, "for", browserType.name());
  })
  // Close browser instance.
  // await browserServer.close();
})();
```

### Motivation

Improve API usability and playability.

### PR DESCRIPTION
**Pull request by cavivie**
Closes [30998](https://github.com/microsoft/playwright/issues/30998).

### CODE REVIEW

File: docs/src/api/class-android.md [hunk 1]
```diff
@@ -202,6 +202,12 @@ Prevents automatic playwright driver installation on attach. Assumes that the dr
 Optional device serial number to launch the browser on. If not specified, it will
 throw if multiple devices are connected.
 
+### option: Android.launchServer.host
+* since: v1.44.2
```
Comment (mxschmitt) [inline docs/src/api/class-android.md:206]: I 
```suggestion
* since: v1.45
```

File: docs/src/api/class-android.md [hunk 2]
```diff
@@ -202,6 +202,12 @@ Prevents automatic playwright driver installation on attach. Assumes that the dr
...
+- `host` <[string]>
+
+Host to use for the web socket. Defaults to localhost that accessible locally.
```
Comment (mxschmitt) [inline docs/src/api/class-android.md:209]:
```suggestion
Host to use for the web socket. Defaults to `localhost`.
```

File: docs/src/api/class-browsertype.md [hunk 3]
```diff
@@ -380,6 +380,12 @@ const { chromium } = require('playwright');  // Or 'webkit' or 'firefox'.
 ### option: BrowserType.launchServer.logger = %%-browser-option-logger-%%
 * since: v1.8
 
+### option: BrowserType.launchServer.host
+* since: v1.44.2
```
Comment (mxschmitt) [inline docs/src/api/class-browsertype.md:384]:
```suggestion
* since: v1.45
```

File: docs/src/api/class-browsertype.md [hunk 4]
```diff
@@ -380,6 +380,12 @@ const { chromium } = require('playwright');  // Or 'webkit' or 'firefox'.
...
+- `host` <[string]>
+
+Host to use for the web socket. Defaults to localhost that accessible locally.
```
Comment (mxschmitt) [inline docs/src/api/class-browsertype.md:387]:
```suggestion
Host to use for the web socket. Defaults to localhost that accessible locally.
```
```suggestion
Host to use for the web socket. Defaults to `localhost`.
```

File: tests/library/browsertype-launch-server.spec.ts [hunk 5]
```diff
@@ -26,6 +26,13 @@ it.describe('launch server', () => {
     await browserServer.close();
   });
 
+  it('should work with host', async ({ browserType }) => {
+    const host = '0.0.0.0';
```
Comment (mxschmitt) [inline tests/library/browsertype-launch-server.spec.ts:30]: Add a test for Android.launchServer as well? Should be very similar.

**Review by mxschmitt — CHANGES_REQUESTED** [review]
LGTM % comments

**Comment by cavivie** [comment]
@mxschmitt Is this dirty error caused by me pushing twice quickly?

**Comment by mxschmitt** [comment]
> @mxschmitt Is this dirty error caused by me pushing twice quickly?

You need to run npm run build, it will then update some generated files.

**Comment by cavivie** [comment]
BTW, I found that although the ws server pointed out that the ws url was `localhost`, it actually listened at 0.0.0.0. This is because the http server listen uses net.Server listen. When the host is omitted in this method, the unspecified address will be used.

https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/utils/wsServer.ts#L57-L83

https://nodejs.org/api/net.html#serverlistenport-host-backlog-callback

To fix it, we need to do either of these things:
1. We may need to open a new bug issue for this and fix it: set hostname to localhost instead of undefined, because undefined means that the net server listens at the unspecified address ipv4(0.0.0.0)/ipv6(::). 
2. We only changed the document description to unspecified, and changed the hostname in the ws url to undefined value to 0.0.0.0 but not`localhost`, [line to change](https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/utils/wsServer.ts#L80).

File: packages/playwright-core/src/utils/wsServer.ts [hunk 6]
```diff
@@ -77,7 +77,7 @@ export class WSServer {
           reject(new Error('Could not bind server socket'));
           return;
         }
-        const wsEndpoint = typeof address === 'string' ? `${address}${path}` : `ws://${hostname || 'localhost'}:${address.port}${path}`;
+        const wsEndpoint = typeof address === 'string' ? `${address}${path}` : `ws://${hostname || address.address}:${address.port}${path}`;
```
Comment (cavivie) [inline packages/playwright-core/src/utils/wsServer.ts:80]: For compatibility, I only changed the `host` option document to unspecified and specified the output ws hostname to be the address output by the server.

**Comment by cavivie** [comment]
> 2\. We only change the document description to unspecified, ~and chang the hostname in the ws url to undefined value to 0.0.0.0 but not`localhost`, [line to change](https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/utils/wsServer.ts#L80)~.

Finally, I followed this to keep compatibility, what do you think?

File: docs/src/api/class-android.md [hunk 7]
```diff
@@ -202,6 +202,12 @@ Prevents automatic playwright driver installation on attach. Assumes that the dr
 Optional device serial number to launch the browser on. If not specified, it will
 throw if multiple devices are connected.
 
+### option: Android.launchServer.host
+* since: v1.5
+- `host` <[string]>
+
+Host to use for the web socket. Defaults to `unspecified`.
```
Comment (pavelfeldman) [inline docs/src/api/class-android.md:209]: Your change looks good. Please remove the mentions of 'unspecified', we imply that for the options, they are optional.

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**